### DISCUSSION TITLE
[@types/inquirer] How should I import type of Inquirer now?

### DISCUSSION THREAD

**Discussion by tal-rofe** [DefinitelyTyped/DefinitelyTyped#65226] [Issues with a @types package] [answered]
When using `@types/inquirer@8.2.1` I could do: `import type { Inquirer } from 'inquirer';`.

But now, when upgrading to `@types/inquirer/9.1.5` I'm getting an error: 
```
Module '"inquirer"' has no exported member 'Inquirer'. Did you mean to use 'import Inquirer from "inquirer"' instead?ts(2614)
```

Well, I tried as suggested to do `import type Inquirer from "inquirer"`, but then I got this error:
```
Cannot use namespace 'Inquirer' as a type.ts(2709)
```

So how can I use the same type as of version 8.2.1, in version 9.1.5?

[a comment by an automated account is not shown]

**Comment by neurolag** [comment, marked as answer]
Sadly, I do not remember what change forced me to remove `Inquirer` from the type declarations.
However, you might want to do something like this:

```ts
let x: typeof import("inquirer").default;
```

or

```ts
import type inquirer from "inquirer";

let x: typeof inquirer;
```

**Reply by tal-rofe** [reply]
Well, did:
`import type * as Inquirer from 'inquirer';`

then ...

`let x: typeof Inquirer.default`

kind of strange though, but ok.

**Reply by neurolag** [reply]
This alternative might look less strange to you:
```ts
import type inquirer from 'inquirer';
let x: typeof inquirer;
```

**Comment by crimson-med** [comment]
Hello, since the update i'm getting multiple issues with using types inquirer.

How it worked before upgrading: 

```ts
import * as inquirer from 'inquirer'

return inquirer.prompt([{
    type: 'list',
    name: 'name',
    message: "What is the city's name?",
    choices: getRandomElement(Data.Cities, 4),
}])
```

but this says there is no defined type prompt.

I have also tried the fixed suggested:

```ts
import type inquirer from "inquirer";
let x: typeof inquirer;
```

Result: 

```
    return x.prompt([
             ^
TypeError: Cannot read properties of undefined (reading 'prompt')

```

**Reply by neurolag** [reply]
Sorry for the late response.
You are probably confusing something as this discussion is on how to import the TYPE of inquirer rather than inquirer itself.

But first, let's just get out the answer in short:
You would have to use a code like this:

```ts
import inquirer from 'inquirer'

// [...]
    return inquirer.prompt([{
        type: 'list',
        name: 'name',
        message: "What is the city's name?",
        choices: getRandomElement(Data.Cities, 4),
    }])
// [...]
```

Next to the issue of the first piece of code:
Since `inquirer@9`, inquirer is `default`-exported. The statement you have to use for `default` exports is the following:

```ts
import varNameOfYourChoice from "inquirer";
```

This is called a default import and it can also be combined with named imports:

```ts
import inquirer, { createPromptModule } from "inquirer";
```

However, a `import * as inquirer from "inquirer";`-statement will stuff all members exported by `"inquirer"` (including the `default` export) into one object.
If you were to use such a star-import, your code would have to look like this:

```ts
import * as inquirer from "inquirer";

inquirer.default.prompt({ /* Whatever prompts */ });
// Other members exported by `"inquirer"` can also be used:
inquirer.createPromptModule();
```

However, I would definitely recommend you with a default export.

The second piece of code is a so-called type import (notice the `type` portion in the import statement):

```ts
import type inquirer from "inquirer";
```

This statement won't actually import `"inquirer"` but rather fetch its type declaration for the use with TypeScript's IntelliSense, autocompletion and type checking.

The second statement, `let x: typeof inquirer;` will just tell TypeScript that there is a variable called `x` that has the same type that the type-imported `inquirer` has. However, because no value is assigned to the variable `x` (using the `=` operation such as `x = {}`), `x` remains `undefined` which causes an error when you try to call `x.prompt()`.

Type imports can be very useful if you, for example, wanted to write a function which accepts an `inquirer`-object but don't want to actually import `"inquirer"`:

```ts
import type inquirer from "inquirer";

export function askMyQuestions(inquirerInstance: typeof inquirer) {
    inquirerInstance.prompt([{
        type: 'list',
        name: 'name',
        message: "What is the city's name?",
        choices: getRandomElement(Data.Cities, 4),
    }])
}
```

This statement causes `askMyQuestions` to force that the first parameter, `inquirerInstance` must have the type of `inquirer`.

This is useful for forcing correct usage of the function:

```ts
import inquirer from "inquirer";
import { askMyQuestions } from "./askMyQuestions.js";

// TypeScript will show errors for these statements
askMyQuestions(console);
askMyQuestions(process);

// But not here
askMyQuestions(inquirer);
```
